File: app/src/main/java/cz/johnyapps/cheers/fragments/BeverageCategoryFragment.java
package cz.johnyapps.cheers.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.material.bottomsheet.BottomSheetBehavior;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import cz.johnyapps.cheers.R;
import cz.johnyapps.cheers.adapters.recycleradapters.CountersAdapter;
import cz.johnyapps.cheers.entities.BeverageCategory;
import cz.johnyapps.cheers.entities.CounterWithBeverage;
import cz.johnyapps.cheers.tools.Logger;
import cz.johnyapps.cheers.tools.SharedPrefsUtils;
import cz.johnyapps.cheers.viewmodels.MainViewModel;
import cz.johnyapps.cheers.views.CustomCoordinatorLayout;
import cz.johnyapps.cheers.views.EmptyMessageRecyclerView;

public class BeverageCategoryFragment extends Fragment implements BackOptionFragment {
    private static final String TAG = "BeverageCategoryFragment";

    @NonNull
    private final BeverageCategory beverageCategory;
    @Nullable
    private SharedPreferences generalPrefs;
    @Nullable
    private BottomSheetBehavior<View> bottomSheetBehavior;
    @Nullable
    private CountersAdapter adapter;
    private long selectedCounterId = -1;

    private MainViewModel viewModel;
    private AppCompatImageView beverageCategoryImageView;

    public BeverageCategoryFragment(@NonNull BeverageCategory beverageCategory) {
        this.beverageCategory = beverageCategory;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setupViewModel();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        CustomCoordinatorLayout root = (CustomCoordinatorLayout) inflater.inflate(R.layout.fragment_beverage_category, container, false);

        selectedCounterId = getGeneralPrefs().getLong(beverageCategory.getSelectedCounterPrefName(), -1);

        setupCategoryImage(root);
        setupBottomSheet(root);
        setupCounterRecyclerView(root);
        setupObservers();
        changePeekHeight(viewModel.getCounterHeight());

        return root;
    }

    @Override
    public boolean onBackPressed() {
        if (isBottomSheetExpanded() && bottomSheetBehavior != null) {
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
            return true;
        }

        return false;
    }

    private void setupCounterRecyclerView(@NonNull View root) {
        EmptyMessageRecyclerView counterRecyclerView = root.findViewById(R.id.countersRecyclerView);
        LinearLayoutManager layoutManager = new LinearLayoutManager(root.getContext());
        counterRecyclerView.setLayoutManager(layoutManager);
        counterRecyclerView.setAdapter(adapter);
        counterRecyclerView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
            if (bottomSheetBehavior != null) {
                bottomSheetBehavior.setDraggable(firstVisibleItemPosition == 0);
            }
        });
    }

    private void setupCategoryImage(@NonNull View root) {
        beverageCategoryImageView = root.findViewById(R.id.beverageCategoryImageView);
        beverageCategoryImageView.setImageDrawable(ContextCompat.getDrawable(root.getContext(), beverageCategory.getImageResId()));
        beverageCategoryImageView.setOnClickListener(v -> playSound(v.getContext()));
    }

    private void setupObservers() {
        viewModel.getCountersWithBeverages().observe(getViewLifecycleOwner(), counterWithBeverages -> {
            counterWithBeverages = new ArrayList<>(counterWithBeverages);

            if (adapter != null) {
                adapter.update(counterWithBeverages);
            }

            if (bottomSheetBehavior != null) {
                changePeekHeight(counterWithBeverages.isEmpty() ? 0 : viewModel.getCounterHeight());
            }

            showSelectedCounter();
        });
    }

    private void setupBottomSheet(@NonNull View root) {
        View bottomSheet = root.findViewById(R.id.bottomSheet);
        bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet);
        bottomSheetBehavior.addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                if (newState == BottomSheetBehavior.STATE_EXPANDED || newState == BottomSheetBehavior.STATE_HALF_EXPANDED) {
                    if (adapter != null) {
                        adapter.setAllCountersDisabled(true);
                    }
                } else {
                    if (adapter != null) {
                        adapter.setAllCountersDisabled(false);
                    }
                }
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {

            }
        });

        adapter = new CountersAdapter(root.getContext(), viewModel.getCountersWithBeverages().getValue());
        adapter.setAllowSelection(false);
        adapter.setOnCounterClickListener(this::selectCounter);
        adapter.setOnSizeChangedListener((width, height) -> changePeekHeight(height));

        if (adapter.isAllCountersDisabled()) {
            adapter.setAllCountersDisabled(false);
        }
    }

    private void showSelectedCounter() {
        List<CounterWithBeverage> counters = viewModel.getCountersWithBeverages().getValue();

        if (counters != null) {
            CounterWithBeverage selectedCounter = null;
            int position = -1;

            if (selectedCounterId > 0) {
                for (int i = 0; i < counters.size(); i++) {
                    if (counters.get(i).getCounter().getId() == selectedCounterId) {
                        position = i;
                        selectedCounter = counters.get(i);
                        break;
                    }
                }
            }

            selectCounter(selectedCounter, position);
        }
    }

    private void changePeekHeight(int peekHeight) {
        viewModel.setCounterHeight(peekHeight);

        if (bottomSheetBehavior != null) {
            bottomSheetBehavior.setPeekHeight(peekHeight);
        }

        ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) beverageCategoryImageView.getLayoutParams();
        layoutParams.setMargins(0, 0, 0, peekHeight);
        beverageCategoryImageView.setLayoutParams(layoutParams);
        beverageCategoryImageView.forceLayout();
    }

    public void selectLastCounter() {
        if (adapter != null) {
            int position = adapter.getItemCount() - 1;
            selectCounter(adapter.getItem(position), position);
        } else {
            Logger.w(TAG, "selectLastCounter: Adapter is null");
        }
    }

    private void selectCounter(@Nullable CounterWithBeverage counterWithBeverage, int position) {
        if (adapter != null) {
            adapter.moveToTop(position);
        }

        selectedCounterId = counterWithBeverage == null ? -1 : counterWithBeverage.getCounter().getId();

        if (bottomSheetBehavior != null) {
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
        } else {
            Logger.w(TAG, "selectCounter: BottomSheetBehavior is null");
        }

        getGeneralPrefs().edit()
                .putLong(beverageCategory.getSelectedCounterPrefName(), selectedCounterId)
                .apply();
    }

    private void setupViewModel() {
        ViewModelProvider provider = new ViewModelProvider(requireActivity());
        viewModel = provider.get(MainViewModel.class);
    }

    private boolean isBottomSheetExpanded() {
        if (bottomSheetBehavior == null) {
            return false;
        }

        int state = bottomSheetBehavior.getState();
        return state == BottomSheetBehavior.STATE_EXPANDED || state == BottomSheetBehavior.STATE_HALF_EXPANDED;
    }

    private void playSound(@NonNull Context context) {
        int[] sounds = beverageCategory.getSounds();

        if (sounds.length == 0) {
            return;
        }

        int soundToPlay;

        if (sounds.length > 1) {
            soundToPlay = sounds[new Random().nextInt(sounds.length)];
        } else {
            soundToPlay = sounds[0];
        }

        MediaPlayer player = MediaPlayer.create(context, soundToPlay);
        player.start();
    }

    @NonNull
    public SharedPreferences getGeneralPrefs() {
        if (generalPrefs == null) {
            generalPrefs = SharedPrefsUtils.getGeneralPrefs(requireContext());
        }

        return generalPrefs;
    }
}

File: app/src/main/java/cz/johnyapps/cheers/fragments/BeverageCategoryFragment.kt
package cz.johnyapps.cheers.fragments

import android.content.Context
import android.media.MediaPlayer
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import cz.johnyapps.cheers.R
import cz.johnyapps.cheers.SharedPrefsNames
import cz.johnyapps.cheers.adapters.recycler.CountersAdapter
import cz.johnyapps.cheers.databinding.FragmentBeverageCategoryBinding
import cz.johnyapps.cheers.entities.BeverageCategory
import cz.johnyapps.cheers.entities.CounterWithBeverage
import cz.johnyapps.cheers.tools.Logger
import cz.johnyapps.cheers.tools.SharedPrefsUtils
import cz.johnyapps.cheers.viewmodels.BeverageCategoriesViewModel
import java.util.*
import kotlin.collections.ArrayList

class BeverageCategoryFragment(): Fragment(), BackOptionFragment {
    private lateinit var binding: FragmentBeverageCategoryBinding
    private lateinit var adapter: CountersAdapter
    private lateinit var bottomSheetBehavior: BottomSheetBehavior<FrameLayout>
    private lateinit var viewModel: BeverageCategoriesViewModel
    private var peekHeightIfCounters = 0
    private var beverageCategory = BeverageCategory.BEER

    constructor(beverageCategory: BeverageCategory): this() {
        this.beverageCategory = beverageCategory
    }

    companion object {
        private const val TAG = "BeverageCategoryFragment"
        private const val BEVERAGE_CATEGORY = "beverage_category"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val name = savedInstanceState?.getString(BEVERAGE_CATEGORY)

        if (name != null) {
            Logger.d(TAG, "onCreate: Setting category to: $name")
            beverageCategory = BeverageCategory.valueOf(name)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(BEVERAGE_CATEGORY, beverageCategory.name)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val provider = ViewModelProvider(requireActivity())
        viewModel = provider.get(BeverageCategoriesViewModel::class.java)

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_beverage_category, container, false)
        val context = binding.root.context

        peekHeightIfCounters = SharedPrefsUtils.getGeneralPrefs(context).getInt(SharedPrefsNames.COUNTER_HEIGHT, 0)

        binding.beverageCategoryImageView.setImageDrawable(ContextCompat.getDrawable(context, beverageCategory.imageResId))
        binding.beverageCategoryImageView.setOnClickListener { playSound(context) }

        setupCountersRecyclerView()
        setupBottomSheetBehavior()
        changePeekHeight(0)
        setupObservers()

        return binding.root
    }

    private fun changePeekHeight(peekHeight: Int) {
        bottomSheetBehavior.peekHeight = peekHeight

        val layoutParams = binding.beverageCategoryImageView.layoutParams as MarginLayoutParams
        layoutParams.setMargins(0, 0, 0, peekHeight)
        binding.beverageCategoryImageView.layoutParams = layoutParams
        binding.beverageCategoryImageView.forceLayout()
    }

    private fun setupCountersRecyclerView() {
        val context = binding.root.context

        adapter = CountersAdapter()
        adapter.setAllowSelection(false)
        adapter.onCounterClickListener = object : CountersAdapter.OnCounterClickListener {
            override fun onClick(counterWithBeverage: CounterWithBeverage) {
                selectCounter(counterWithBeverage)
            }
        }

        binding.countersRecyclerView.layoutManager = LinearLayoutManager(context)
        binding.countersRecyclerView.adapter = adapter
    }

    private fun setupBottomSheetBehavior() {
        bottomSheetBehavior = BottomSheetBehavior.from(binding.bottomSheet)
        bottomSheetBehavior.addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                adapter.allCountersDisabled = newState == BottomSheetBehavior.STATE_EXPANDED ||
                        newState == BottomSheetBehavior.STATE_HALF_EXPANDED
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {

            }
        })
    }

    fun selectCounter(counterWithBeverage: CounterWithBeverage) {
        Logger.d(TAG, "selectCounter: Selected counter: ${counterWithBeverage.beverage.name}")

        binding.countersRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                //For some reason it has to be there
            }
        })
        binding.countersRecyclerView.scrollToPosition(0)

        val context = binding.root.context
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED

        val selectedCounterId = counterWithBeverage.counter.id
        SharedPrefsUtils.getGeneralPrefs(context).edit()
            .putLong(beverageCategory.selectedCounterPrefName, selectedCounterId)
            .apply()

        submitList(viewModel.countersWithBeverages.value)
    }

    private fun submitList(it: List<CounterWithBeverage>?) {
        if (it != null) {
            val selected = SharedPrefsUtils.getGeneralPrefs(binding.root.context).getLong(beverageCategory.selectedCounterPrefName, -1)
            val list = ArrayList(it) as MutableList<CounterWithBeverage>

            if (selected > -1) {
                for (i in 0 until list.size) {
                    val item = list[i]
                    if (item.counter.id == selected) {
                        list.removeAt(i)
                        list.add(0, item)
                        break
                    }
                }
            }

            adapter.submitList(list)
        } else {
            adapter.submitList(null)
        }
    }

    private fun setupObservers() {
        viewModel.countersWithBeverages.observe(viewLifecycleOwner, {
            if (it.isNotEmpty()) {
                changePeekHeight(peekHeightIfCounters)
            } else {
                changePeekHeight(0)
            }

            submitList(it)
        })
    }

    private fun playSound(context: Context) {
        val sounds = beverageCategory.sounds

        if (sounds.isEmpty()) {
            return
        }

        val soundToPlay: Int = if (sounds.size > 1) {
            sounds[Random().nextInt(sounds.size)]
        } else {
            sounds[0]
        }

        val player = MediaPlayer.create(context, soundToPlay)
        player.start()
    }

    override fun onBackPressed(): Boolean {
        if (bottomSheetBehavior.state == BottomSheetBehavior.STATE_EXPANDED ||
            bottomSheetBehavior.state == BottomSheetBehavior.STATE_HALF_EXPANDED) {
            bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
            return true
        }

        return false
    }
}